backendRoot/src/main/java/com/virtuslab/gitmachete/backend/root/GitMacheteRepositoryBuilder.java

```diff
@@ -241,10 +269,23 @@ private SyncToParentStatus deriveSyncToParentStatus(IGitCoreRepository gitCoreRe
     }
   }
 
-  /**
-   * @return Option of {@link IGitCoreLocalBranch} or if branch with given name doesn't exist returns empty Option
-   */
-  private Optional<IGitCoreLocalBranch> getCoreBranchFromName(IGitCoreRepository gitCoreRepository, String branchName) {
-    return Try.of(() -> Optional.of(gitCoreRepository.getLocalBranch(branchName))).getOrElse(() -> Optional.empty());
+  @RequiredArgsConstructor
+  private static class AncestorityCache {
+    private final IGitCoreRepository repository;
+    Map<Tuple2<BaseGitCoreCommit, BaseGitCoreCommit>, Boolean> cache = new HashMap<>();
+
+    boolean isAncestor(BaseGitCoreCommit presumedAncestor, BaseGitCoreCommit presumedDescendant)
+        throws GitMacheteException {
+      Tuple2<BaseGitCoreCommit, BaseGitCoreCommit> key = Tuple.of(presumedAncestor, presumedDescendant);
+      Boolean isAncestorResult = cache.get(key);
+      if (isAncestorResult != null) {
```

Comment:
I know this is WIP but while we are here just small optimization: condition can be changed to `isAncestorResult == null` and then`return isAncestorResult` as a common part can be extracted from `if`

Reply:
good point!